### PR TITLE
CClient: adding core API tests.

### PR DESCRIPTION
**Pull request by oopsmonk**
Adding Core API test cases using devnet.

- [x] test_check_consistency
- [x] test_find_transactions
- [x] test_get_balances
- [x] test_get_inclusion_states
- [x] test_get_node_info
- [x] test_get_tips
- [x] test_get_trytes
- [x] test_were_addresses_spent_from
- [x] test_add_neighbors
- [x] test_attach_to_tangle
- [x] test_broadcast_transactions
- [x] test_get_neighbors
- [x] test_remove_neighbors
- [x] test_store_transactions

### CODE REVIEW

File: cclient/serialization/json/get_inclusion_states.c [hunk 1]
```diff
@@ -13,6 +13,12 @@ retcode_t json_get_inclusion_states_serialize_request(get_inclusion_states_req_t
   retcode_t ret = RC_ERROR;
   char const *json_text = NULL;
   log_debug(json_logger_id, "[%s:%d]\n", __func__, __LINE__);
+
+  if (!req->tips && !req->transactions) {
```
Comment (thibault-martinez) [inline cclient/serialization/json/get_inclusion_states.c]: Aren't `tips` optional ? 
https://docs.iota.org/docs/iri/0.1/references/api-reference#getinclusionstates

File: cclient/response/get_neighbors.c [hunk 2]
```diff
@@ -33,7 +33,10 @@ get_neighbors_res_t* get_neighbors_res_new() {
   return nbors;
 }
 
-void get_neighbors_res_free(get_neighbors_res_t* res) { utarray_free(res); }
+void get_neighbors_res_free(get_neighbors_res_t** res) {
+  utarray_free(*res);
```
Comment (thibault-martinez) [inline cclient/response/get_neighbors.c:41]: Check before dereferencing

File: cclient/response/get_inclusion_states.c [hunk 3]
```diff
@@ -34,7 +34,12 @@ bool get_inclusion_states_res_states_at(get_inclusion_states_res_t* in, size_t i
   return false;
 }
 
-size_t get_inclusion_states_res_states_count(get_inclusion_states_res_t* in) { return utarray_len(in->states); }
+size_t get_inclusion_states_res_states_count(get_inclusion_states_res_t* in) {
+  if (!in->states) {
```
Comment (thibault-martinez) [inline cclient/response/get_inclusion_states.c:50]: Check `in` before accessing `states`

File: cclient/response/attach_to_tangle.c [hunk 4]
```diff
@@ -30,7 +30,12 @@ flex_trit_t* attach_to_tangle_res_trytes_at(attach_to_tangle_res_t* res, int ind
   return hash_array_at(res->trytes, index);
 }
 
-size_t attach_to_tangle_res_trytes_cnt(attach_to_tangle_res_t* res) { return hash_array_len(res->trytes); }
+size_t attach_to_tangle_res_trytes_cnt(attach_to_tangle_res_t* res) {
+  if (!res->trytes) {
```
Comment (thibault-martinez) [inline cclient/response/attach_to_tangle.c:50]: Check `res` before accessing `trytes`

File: cclient/api/tests/BUILD [hunk 5]
```diff
@@ -0,0 +1,181 @@
+cc_test(
```
Comment (thibault-martinez) [inline cclient/api/tests/BUILD:1]: No test for gTTA ?

File: cclient/request/add_neighbors.c [hunk 6]
```diff
@@ -28,9 +28,18 @@ retcode_t add_neighbors_req_uris_add(add_neighbors_req_t* req, char const* const
   return RC_OK;
 }
 
-size_t add_neighbors_req_uris_len(add_neighbors_req_t* req) { return utarray_len(req->uris); }
+size_t add_neighbors_req_uris_len(add_neighbors_req_t* req) {
+  if (!req->uris) {
```
Comment (thibault-martinez) [inline cclient/request/add_neighbors.c:40]: Check `req`

File: cclient/request/add_neighbors.c [hunk 7]
```diff
@@ -28,9 +28,18 @@ retcode_t add_neighbors_req_uris_add(add_neighbors_req_t* req, char const* const
...
+    return 0;
+  }
+  return utarray_len(req->uris);
+}
 
 const char* add_neighbors_req_uris_at(add_neighbors_req_t* req, size_t idx) {
+  if (!req->uris) {
```
Comment (thibault-martinez) [inline cclient/request/add_neighbors.c:52]: Check `req`

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: cclient/serialization/json/get_inclusion_states.c (hunk 1 above)
Comment (oopsmonk) [reply cclient/serialization/json/get_inclusion_states.c]: optional, but need the json array.

File: cclient/api/tests/cclient_service_init.c [hunk 8]
```diff
@@ -0,0 +1,46 @@
+#include "cclient/api/tests/cclient_test_defs.h"
```
Comment (thibault-martinez) [inline cclient/api/tests/cclient_service_init.c:8]: Copyright :)

File: cclient/api/tests/test_find_transactions.c [hunk 9]
```diff
@@ -0,0 +1,202 @@
+
```
Comment (thibault-martinez) [inline cclient/api/tests/test_find_transactions.c]:
```suggestion
```

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**